operator/controllers/resources/role_controller_test.go

```diff
@@ -33,57 +39,189 @@ import (
 		if trace.IsNotFound(err) {
 			return false
 		}
 		require.NoError(t, err)
 
+		// Role should have the same name, and have the Kubernetes origin label
 		require.Equal(t, tRole.GetName(), roleName)
-
 		require.Contains(t, tRole.GetMetadata().Labels, types.OriginLabel)
 		require.Equal(t, tRole.GetMetadata().Labels[types.OriginLabel], types.OriginKubernetes)
 
 		return true
 	})
 
-	// The role is deleted in K8S
+	// Cleanup and test, we delete the role in Kubernetes```
 	k8sDeleteRole(ctx, t, k8sClient, roleName, ns.Name)
 
+	// We wait for the role to be deleted in Teleport
 	fastEventually(t, func() bool {
 		_, err := tClient.GetRole(ctx, roleName)
 		return trace.IsNotFound(err)
 	})
 }
 
-func TestRoleDeletionDrift(t *testing.T) {
-	ctx := context.Background()
+func TestRoleCreationFromYAML(t *testing.T) {
+	ctx, tClient, k8sClient, ns := setupKubernetesAndTeleport(t)
+	tests := []struct {
+		name         string
+		roleSpecYAML []byte
+		shouldFail   bool
+		expectedSpec *types.RoleSpecV5
+	}{
+		{
+			"Valid login list",
+			[]byte(`
+allow:
+  logins:
+  - ubuntu
+  - root
+`),
+			false,
+			&types.RoleSpecV5{
+				Allow: types.RoleConditions{
+					Logins: []string{"ubuntu", "root"},
+				},
+			},
+		},
+		{
+			"Valid node_labels wildcard (list version)",
+			[]byte(`
+allow:
+  node_labels:
+    '*': ['*']
+`),
+			false,
+			&types.RoleSpecV5{
+				Allow: types.RoleConditions{
+					NodeLabels: map[string]apiutils.Strings{
+						"*": {"*"},
+					},
+				},
+			},
+		},
+		{
+			"Valid node_labels wildcard (string version)",
+			[]byte(`
+allow:
+  node_labels:
+    '*': '*'
+`),
+			false,
+			&types.RoleSpecV5{
+				Allow: types.RoleConditions{
+					NodeLabels: map[string]apiutils.Strings{
+						"*": {"*"},
+					},
+				},
+			},
+		},
+		{
+			"Invalid node_labels (label value is integer)",
+			[]byte(`
+allow:
+  node_labels:
+    'foo': 1
+`),
+			true,
+			nil,
+		},
+		{
+			"Invalid node_labels (label value is object)",
+			[]byte(`
+allow:
+  node_labels:
+    'foo':
+      'bar': 'baz'
+    'logins':
+      - 'ubuntu'
+`),
+			true,
+			nil,
+		},
+	}
 
-	teleportServer, operatorName := defaultTeleportServiceConfig(t)
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			// Creating the Kubernetes resource. We are using an untyped client to be able to create invalid resources.
+			roleManifest := map[string]interface{}{}
+			err := yaml.Unmarshal(tc.roleSpecYAML, &roleManifest)
+			require.NoError(t, err)
+
+			roleName := validRandomResourceName("role-")
+
+			obj := getUnstructuredObjectFromGVK(TeleportRoleGVK)
+			obj.Object["spec"] = roleManifest
+			obj.SetName(roleName)
+			obj.SetNamespace(ns.Name)
+			err = k8sClient.Create(ctx, obj)
+			require.NoError(t, err)
+
+			// If failure is expected we should not see the resource in Teleport
+			if tc.shouldFail {
+				// We wait 1 second to ensure reconciliation happened
+				time.Sleep(time.Second)
```

Comment:
Can we replace this `Sleep` with an `Eventually`?

Reply:
Not really, because if I understand correctly `Evevntally` will retry gradually. Here I want to assert this does not work after 1 second. If I use eventually it will fail on the first try, maybe before the operator finishes its reconciliation. I think the way to fix this test would be to get the kubernetes object and validate the error condition is here. I'll fix it today.

Reply:
Oh I see.
So, we should ideally wait for the reconciliation to happen :thinking:
Yeah, validating the error condition of the K8s object seems to be a good approach :+1: